### PR TITLE
Fix lgamma fp32 coverage gap and plateau near z=0.5

### LINKED ISSUE #51976: [kernels]: lgamma (fp32) returns one identical value for all 167,773 floats in [0.5, 0.51], then jumps 0.071
### Component / Area

kernels

### Issue Type (optional)

Bad Outputs

### Observed

Caveat, stated up front: I don't have Tenstorrent hardware. This is read from the kernel source and checked against float64 on CPU, not measured on silicon.

`lgamma` is strictly decreasing on (0, 1.46163…), so it cannot be constant on any interval. On the fp32 path it is: every one of the **167,773 float32 values in [0.5f, 0.51f]** returns the identical bit pattern `0x3F128F51` = 0.5723649263381958.

Then, between two *consecutive* float32, it jumps:

```
lgamma(0.50999999046325684)  ->  0.572364926     (bits 0x3F028F5C)
lgamma(0.51000005006790161)  ->  0.501337528     (bits 0x3F028F5D)
                                  ^ a step of 0.0710
```

The true function moves 1.14e-07 between those two inputs.

**Defect 1 — a point patch written as a window patch.**

`ckernel_sfpu_lgamma.h`:

```cpp
 60:  constexpr float LOG_SQRT_PI = 0.57236494f;  // lgamma(0.5) = ln(sqrt(pi))
...
117:  // Handle boundary case
118:  v_if(sfpi::abs(z - 0.5f) < 0.01f) { res = LOG_SQRT_PI; }
```

Both the constant's comment and `// Handle boundary case` say the intent was the *point* z = 0.5. The condition is a *tolerance*. Since `z = (in < 0.5) ? 1 - in : in` is always ≥ 0.5, only the upper half of the window is reachable — z in [0.5, 0.51) — and the whole of it is replaced by the literal.

The boundary is exact in fp32: `0.51f - 0.5f = 0.00999999046325684` (Sterbenz-exact) which is below `0.01f = 0.00999999977648258`, so 0.51f is inside; the next float32 up is not.

Across that interval the true `lgamma` spans [0.552973836, 0.572364943] — a width of 0.0193911 that the kernel compresses to zero. Worst absolute error 0.01939 at x = 0.51f, i.e. 3.51% relative, 325,329 fp32 ULP.

**Defect 2 — the branches do not tile the domain, and this is the underlying cause.**

The Taylor bridges are `v_if(abs_range1 < 0.25f)` (line 87) and `v_elseif(abs_range2 < 0.25f)` (line 97), covering z > 0.75 and z near 2. The plateau covers z < 0.51. **Nothing covers z in [0.51, 0.75)**, so it falls through to the Stirling branch, whose Bernoulli series truncated at r3 = -1/1680 has not converged at those z:

| z | kernel | correct (float64) | rel. error |
|---|---|---|---|
| 0.5100001 | 0.50133741 | 0.55297363 | 9.34% |
| 0.52 | 0.48933682 | 0.53405989 | 8.37% |
| 0.60 | 0.38282782 | 0.39823386 | 3.87% |
| 0.74 | 0.21110904 | 0.21426755 | 1.47% |
| 0.751 | — | — | 0.005% ← Taylor takes over |

The two-orders-of-magnitude drop at the z = 0.75 seam is what identifies this as a coverage gap rather than rounding. By the reflection `z = (x < 0.5) ? 1 - x : x`, it affects x in (0.25, 0.49] and x in [0.51, 0.75) symmetrically. A dense sweep of 19,999 points over (0.001, 5.0) finds 1,996 (9.98%) with absolute error above 1e-3, **all of them inside [0.250213, 0.749888]**; outside [0.24, 0.76] the worst error is 5.2e-05.

Worth noting the direction, because it is the opposite of what it looks like: line 118 is a *mitigation* of defect 2, not a cause. Measured counterfactually, dropping the plateau and letting z = 0.5 fall to Stirling gives an error of 0.0598 instead of 1.7e-08. Removing the line would make things worse; widening the Taylor bridge is what fixes both.

Scope: fp32 (and bfloat8_b / bfloat4_b, which route through it). **bfloat16 is not affected** — `unary_op_utils.cpp` sends BFLOAT16 to `lgamma_fast_kernel.cpp` → `calculate_lgamma_stirling`, a separate variant without this line.

**Relates to #40340**, which proposes replacing this kernel with a LUT-based rational approximation. I am filing this as a defect report rather than a competing proposal: if that replacement lands, these two are the cases worth keeping as acceptance criteria for it — strict monotonicity across (0.25, 0.75), and no interval of the domain mapping to a single value. If it does not land soon, the coverage gap in defect 2 is fixable on its own by widening the existing Taylor bridge.

### Expected

`lgamma` is strictly decreasing on (0, 1.46163…) and strictly increasing after, so no interval of the domain may map to a single value, and adjacent float32 inputs may not produce a 0.071 step.

`torch.lgamma` is the registered golden:

```
torch.lgamma(torch.tensor([0.50, 0.502, 0.505, 0.5075, 0.51], dtype=torch.float64))
-> tensor([0.5723649429, 0.5684477701, 0.5626087297, 0.5577762385, 0.5529738362])
```

A suggested direction rather than a demand: widening the Taylor bridge around z = 1 from `0.25f` to `0.5f` would cover z down to 0.5 and close both defects at once — the 5-term Taylor already gives 5e-05 at z = 0.751 and degrades smoothly. At that point line 118 becomes unnecessary. A third bridge centred on z = 0.5 would also work. I have not measured which is cheaper in instructions.

### 1. Steps (exact commands)

```python
import torch, ttnn

# the plateau: 167,773 distinct inputs, one output
x = ttnn.from_torch(torch.tensor([0.5, 0.502, 0.505, 0.5075, 0.51]),
                    dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
print(ttnn.to_torch(ttnn.lgamma(x)))
# observed: tensor([0.5723649, 0.5723649, 0.5723649, 0.5723649, 0.5723649])
# expected: tensor([0.5723649, 0.5684478, 0.5626087, 0.5577762, 0.5529738])

# the step, between two consecutive float32
y = ttnn.from_torch(torch.tensor([0.50999999046325684, 0.51000005006790161]),
                    dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
print(ttnn.to_torch(ttnn.lgamma(y)))
# observed: tensor([0.5723649, 0.5013375])   <- 0.071 apart
# expected: tensor([0.5529738, 0.5529737])   <- 1.1e-07 apart

# the coverage gap
z = ttnn.from_torch(torch.tensor([0.52, 0.60, 0.74, 0.40]),
                    dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
print(ttnn.to_torch(ttnn.lgamma(z)))
# observed: tensor([0.4893368, 0.3828278, 0.2111090, ...])
# expected: tensor([0.5340599, 0.3982339, 0.2142676, ...])
```

Golden: `torch.lgamma(x.double()).float()`.

A monotonicity assertion also catches it without any golden: `lgamma` must be strictly decreasing below 1.46163, and over 1000 steps in [0.5, 0.51] the kernel violates that 1000 times while float64 violates it 0 times.

### 2. Input data / link or description

No dataset needed. Any float32 tensor with a value in [0.5, 0.51] shows the plateau; any value in (0.25, 0.75) excluding that window shows the coverage gap. All are ordinary finite float32 values.

The plateau was confirmed exhaustively over the full bit range 0x3F000000 … 0x3F028F5C (167,773 values); the coverage gap over a dense sweep of 19,999 points in (0.001, 5.0).

### 3. Frequency

Always. Pure arithmetic with no timing, ordering or state dependence.

### 1. Software Versions

tt-metal: main as of 2026-08-02
Reference environment for the golden values: Python 3.14.5 + numpy, math.lgamma in float64, x86-64 Windows
OS version: not applicable — see Hardware Details

### 2. Hardware Details

None. I do not have a Tenstorrent device, so I have not executed this on silicon, and I am stating that up front rather than implying otherwise.

The finding is read from `ckernel_sfpu_lgamma.h` and checked against `math.lgamma` in float64 on CPU, with `sfpu_reciprocal_iter<2>` emulated bit-for-bit from `ckernel_sfpu_recip.h` (validated to 8.8e-08 relative). The emulation reproduces the kernel to 0–1 fp32 ULP where the kernel is expected to be right (x = 3, 5, 10, 20, 50, 100, 1000, the Stirling branch), which is what gives me confidence in the numbers where it is wrong. Substituting an exact reciprocal for the emulated one changes none of the reported errors, so the defect is series truncation rather than fp32 noise.

Happy to be corrected by anyone who can run it.

### Is this a regression?

Unknown

### Regression Details

Not investigated. The plateau and the 0.25f bridges appear to be part of the original structure of this kernel rather than a later change; I did not bisect.

### Logs & Diagnostics

No crash or log output — both defects are silent.

For triage: the existing sweep at `tests/sweep_framework/sweeps/eltwise/unary/lgamma/lgamma.py` draws inputs from `torch_random(low=0.0001, high=100)`. The plateau is 0.01 wide out of a 100-wide range, so roughly 1 sample in 10,000 lands in it, and a correlation-based check absorbs the rest. A monotonicity assertion over a dense grid in (0.25, 0.75) would catch both defects in one test and needs no golden.

### Priority

None

### Impact

Affected workflows: `ttnn.lgamma` on float32, bfloat8_b and bfloat4_b. bfloat16 uses a different kernel and is unaffected.

The practical risk is that both failures are inside (0.25, 0.75), which is a very ordinary range for a log-gamma argument — beta and Dirichlet log-densities, for instance, evaluate there routinely. The errors are up to 9% and entirely silent.

Release or date risk: none known.

### PR DESCRIPTION
**Pull request by patelkrish5090**
Bug fix

Summary

calculate_lgamma_stirling_fp32 (wormhole_b0 and blackhole) had no Taylor bridge covering the range z in [0.5, 0.75). Inputs in that range fell through to the Stirling branch before it had converged, giving up to 9.3% error with no crash or warning.

A boundary patch meant for the single point z=0.5 (abs(z - 0.5f) < 0.01f) was too wide. It made 167,773 different float32 inputs return the same output value.

This PR adds a Taylor bridge centered at z=0.625 to close the gap, and removes the boundary patch since it is not needed anymore. It also adds a boundary_regression suite to the lgamma sweep test (monotonicity check and tight tolerance check over the affected range, float32 dtype), since the existing suites sample too broadly to reliably catch this bug.

Fixes #51976

Notes for reviewers

- The main change is the new Taylor bridge in ckernel_sfpu_lgamma.h (wormhole_b0 and blackhole, same fix applied to both). I checked the new coefficients against mpmath.loggamma at 50 digit precision. Worst relative error over the previously broken range drops from up to 9.3% to about 4.5e-5.
- I compiled calculate_lgamma_stirling_fp32 for both wormhole_b0 and blackhole using the pinned SFPU compiler (sfpi 7.67.0). It compiles cleanly, with only pre-existing warnings unrelated to this change.
- I could not run this on real hardware, since no device was available in my environment. Please run the new sweep suite (boundary_regression in tests/sweep_framework/sweeps/eltwise/unary/lgamma/lgamma.py) on real hardware before merging, and confirm no PCC regression on the existing nightly and xfail suites.

### CODE REVIEW

**Comment by EazyHood** [comment]
Thanks for picking this up — removing the `abs(z - 0.5f) < 0.01f` override is the right call, and the sweep entry that samples only `[0.24, 0.76]` is a good catch on why the existing suite could not see it.

One measurement on the new branch, in case it is useful before this lands.

Evaluating the `z < 0.75f` polynomial in fp32 against `lgamma` at 40 digits, over 4000 points in `[0.5, 0.75)`:

| z | relative error | fp32 ULP |
|---|---|---|
| 0.5 | 2.25e-05 | 216 |
| 0.505 | 1.78e-05 | 168 |
| **0.625** | **1.90e-08** | **0** |
| 0.74 | 2.62e-05 | 377 |
| 0.7499 | 4.46e-05 | **609** |

Worst over the window is **609 ULP at z → 0.75**. It is exact at the expansion point and degrades toward both seams, which is the expected shape for a degree-5 Taylor series covering a half-width of 0.125 — the `z ≈ 1` branch above it uses the same degree over the same half-width and sits in the same place.

Two things that make this worth a look rather than a nit:

- The comment three lines above the coefficient block reads `// Minimal coefficients for 0-3 ULP`. After this change that is no longer true for `z ∈ [0.5, 0.75)`, so either the bound or the comment should move.
- `atol=1e-3, rtol=1e-3` in the new monotonicity check passes with about two orders of magnitude to spare, so the regression suite would not notice if the seam got worse later.

`s0 = 0.36082950` matches `lgamma(0.625) = 0.360829495` to eight digits, so the expansion point is right; it is only the order that is short. Extending to degree 7, or splitting the window at 0.625, brings the seams down to single-digit ULP.

Reproducible without hardware — the branch is pure polynomial arithmetic:

```python
import numpy as np
from mpmath import mp, mpf, loggamma
mp.dps = 40
S = [0.36082950, -1.45270876, 1.70059159, -1.47809690, 1.68212021, -2.11689171]

def poly(z):
    d = np.float32(z) - np.float32(0.625)
    r = np.float32(0)
    for c in reversed(S):
        r = np.float32(r * d + np.float32(c))
    return float(r)

for z in (0.5, 0.625, 0.7499):
    t = float(loggamma(mpf(z)))
    print(z, abs(poly(z) - t) / abs(t))
```

Happy to be wrong on the ULP accounting if you are measuring it differently.

**Comment by patelkrish5090** [comment]
Hey there, thanks for measuring this. I extended the branch to degree 7 (added s6, s7).
Worst case error over [0.5, 0.75) is now about 4e-7, around 9 ULP, down
from about 4.5e-5.

I also tightened the test tolerance from 1e-3 to 1e-4. I checked the
kernel's normal error in the rest of that test window first, so the
tighter number still passes on its own but will catch this kind of
regression next time.

On the "0-3 ULP" comment, I checked and it sits above a different set of
coefficients (used in the Stirling branch for large z), not the code I
added. So I left it as is, but I added a short accuracy note on the new
branch itself so it is clear what to expect there.

**Comment by EazyHood** [comment]
You are right, and thanks for checking rather than taking my word for it — the `0-3 ULP` comment sits above the Stirling coefficients, not the branch you added. My line attributing that bound to the new polynomial was wrong. The accuracy note you put on the branch itself is the better fix.

I re-measured with s6/s7 from the updated diff, same method as before (fp32 Horner, 4000 points in `[0.5, 0.75)`, against `lgamma` at 40 digits):

| | worst absolute | worst ULP | at |
|---|---|---|---|
| degree 5 | `1.289e-05` | 609 | z = 0.50000 |
| **degree 7** | **`4.537e-07`** | **19** | z = 0.50037 |

The absolute figure agrees with your ~4e-7. My ULP count comes out at 19 rather than 9 — that is almost certainly the accounting rather than the polynomial, since the absolute errors match; I flagged in my first comment that I might be counting differently, and this looks like where. Either way it is a ~28x improvement and the seam is no longer the dominant error in the window.

Tightening the test to `1e-4` after checking the rest of the window still passes is the right order to do it in: the assertion now has about two orders of margin instead of four, so it would actually catch a regression at that seam rather than sailing past it.

**Comment by patelkrish5090** [comment]
Sounds good, thanks for double checking the numbers. Since both changes look confirmed on your end, is this ready to merge, or is there anything else you want changed first?

**Comment by EazyHood** [comment]
> Sounds good, thanks for double checking the numbers. Since both changes look confirmed on your end, is this ready to merge, or is there anything else you want changed first?



I am not a maintainer here, so I cannot approve or merge — but from my side both points are
closed: the degree-7 branch measures at 4.5e-07 worst absolute over `[0.5, 0.75)`, and moving the
accuracy note onto the new branch was the right call rather than touching the Stirling comment.

Two things for whoever picks up the review.

The branch is behind `main` and GitHub reports `mergeable_state: behind`, so it will need an
update before it can go in.

The second is worth knowing before anyone assumes CI covers this. I tried to measure your change
on the actual kernel under ttsim rather than by evaluating the polynomial in software, and got
byte-identical results before and after — which turned out to be the measurement, not the patch.
The LLK harness routes `MathOperation.Lgamma` to `calculate_lgamma_stirling`:

```
tt_metal/tt-llk/tests/helpers/include/sfpu_operations.h:1307
    SFPU_UNARY_CALL(..., calculate_lgamma_stirling, ...)
```

`calculate_lgamma_stirling_fp32`, the function this PR changes, is reached only from
`ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/lgamma_kernel.cpp:72` via
`lgamma_stirling_float_tile(0, 1, 0)`, which is a three-operand tile op the LLK suite does not
drive. So the whole `tt_metal/tt-llk/tests` suite passes this change either way, and the sweep
entry you added is the only thing that actually exercises it.

Not a problem with the PR — the sweep is the right place for it — but it does mean a green LLK
run says nothing here, and I would rather flag that than have it read as coverage.

---

**Review by dchenTT — APPROVED** [review]

**Comment by EazyHood** [comment]
You're right about the `0-3 ULP` comment — it sits above the Stirling coefficients, not the branch you added. My mistake.

On the degree-7 branch I get a higher worst case than 9 ULP. Sweeping the last few hundred float32 below the seam:

| z | rel. error | fp32 ULP |
|---|---|---|
| 0.5 | 6.54e-07 | 6 |
| 0.505 | 4.43e-07 | 4 |
| 0.625 | 1.90e-08 | 0 |
| 0.74 | 6.69e-07 | 10 |
| **0.749975** | **1.43e-06** | **20** |

Worst is 20 ULP as `z → 0.75`, not ~9. A uniform grid over `[0.5, 0.75)` misses it: the error climbs steeply only in the last sliver before the seam, so the sample has to be denser there — or just walk `nextafter` down from `0.75f`.

Only flagging it because the new comment reads `accurate to < 10 ULP`. Either the bound moves to ~20, or splitting the window at 0.6875 keeps it under 10.

ULP here is `|poly(z) - lgamma(z)| / spacing(float32(lgamma(z)))`, reference from mpmath at 40 digits, evaluating your s0..s7 in fp32 with Horner. Happy to compare if you measure it differently.


Ready to merge from my side — nothing else blocking.

One optional wording note, take it or leave it. The in-code comment reads
`accurate to < 10 ULP`, and our two ULP counts differ (9 vs 19) while the absolute
error agrees (~4-5e-7). Stating it as absolute error instead — say
`accurate to ~5e-7 over [0.5, 0.75)` — is true under either accounting and
cannot go stale if someone measures ULP a third way.

Not worth another round trip if you would rather ship it as is.
